src/core/model.jl

```diff
@@ -96,6 +98,39 @@ function Base.show(io::IO, abm::ABM{A}) where {A}
     end
 end
 
+"""
+    agent_validator(agent, space)
+Validate the user supplied agent (subtype of `AbstractAgent`).
+Checks for mutability and existance and correct types for fields depending on `SpaceType`.
+"""
+function agent_validator(::Type{A}, space::S) where {A<:AbstractAgent, S<:SpaceType}
+    # Check A for required properties & fields
+    isbitstype(A) && throw(ArgumentError("Agent struct must be mutable"))
```

Comment:
Although technically correct, it is better to use `isimmutable(A)` here.

Reply:
I've done it this way since `isimmutable` only takes `::DataType`. When users provide composite types (for example `Cell` in CA1D.jl), isimmutable ends up testing against `UnionAll` and returns true on a mutable struct.

Is there a nicer way to test this without having to make `Cell` a concrete type like `Cell{Integer, String}`?

Reply:
Providing a Union type is incorrect and should not be accepted because it reduces performane by orders of magnitude.. I am fine with throwing an extra error on union types, saying "please give concrete type" or something.

The code in CA1D.jl should be changed as well. I didn't realize it before.

Reply:
This is something important and not discussed so far in tutorial.md! Thanks for pointing it out! Althought by throwing the error here we won't have to discuss it explicitly

Reply:
OK, great. I'll add a concrete type check and swap this out.

Reply:
@kavir1698 please have a look at this discussion. I often see you using the same code, but its type unstable. If you define `mutable struct Agent{D, F<:AbstractFloat}` and then you give `Agent` to `ABM`, this will create a dictioanry with value type `Agent`. But `Agent` is not a concrete type, this is type instability, and therefore bad for performance.

Reply:
@Datseris I thought this is the recommended way to define types in [the Julia docs](https://docs.julialang.org/en/v1/manual/performance-tips/index.html#Avoid-fields-with-abstract-type-1).

Reply:
Interesting discussion. Perhaps we should run some benchmarks on both of these implementations.

Reply:
@Datseris could you clarify a bit here? the Julia manual clearly states contrary to what you suggest.

Reply:
Let's continue in #171, I think I know what happened here.

Reply:
Guys, no, there is no reason to run any benchmarks, this is basic Julia knowledge.

Yes, `struct Agent{D, F<:AbstractFloat}` is the correct way to define a parameteric type, but `Agent` is not a concrete type. `Agent{Int, Float64}` is.

Therefore `Dict{Int, Agent}` is NOT a concrete type either. `Dict{Int, Agent{Int, FLoat64}}` is. This is not an opinion, this is the truth of the type system.